graph/path/bellman_ford_moore.go

```diff
@@ -27,10 +30,63 @@ func BellmanFordFrom(u graph.Node, g graph.Graph) (path Shortest, ok bool) {
 	path = newShortestFrom(u, nodes)
 	path.dist[path.indexOf[u.ID()]] = 0
 
-	// TODO(kortschak): Consider adding further optimisations
-	// from http://arxiv.org/abs/1111.5414.
-	for i := 1; i < len(nodes); i++ {
-		changed := false
+	// queue to keep track which nodes need to be relaxed
+	// only nodes whose vertex distance changed in the previous iterations need to be relaxed again
+	queue := newBellmanFordQueue()
```

Comment:
I would keep a `bellmanFordQueue`, but have it hold the `[]bool` and the `map[int64]int` that `path` created. This way all the handling of `onQueue` can be dealt with there.

```
type bellmanFordQueue struct {
    queue linear.NodeQueue

    onQueue []bool
    indexOf  map[int64]int
}

func newBellmanFordQueue(indexOf map[int64]int) bellmanFordQueue {
    return bellmanFordQueue{onQueue; make([]bool, len(indexOf), indexOf: indexOf}
}

func (q bellmanFordQueue) enqueue(n graph.Node) {
    i := q.indexOf[n.ID()]
    if q.onQueue[i] {
        panic("bellman-ford: already queued")
    }
    q.onQueue[i] = true
    q.queue.Enqueue(n)
}

func (q bellmanFordQueue) dequeue() graph.Node {
    n := q.queue.Dequeue()
    q.onQueue[q.indexOf[n.ID()]] = false
    return n
}

func (q bellmanFordQueue) has(id int64) bool { return q.onQueue[q.indexOf[id]] }

func (q bellmanFordQueue) len() int { return q.queue.Len() }
```

Reply:
Done! Thank you very much! :)